File: python/section4/q34.py
from q30 import load_neko
import pandas as pd
from pprint import pprint

neko = load_neko()
noun_no_noun = set()
for sentence in neko:
    morphs = pd.DataFrame(sentence)
    noun_idx = morphs[morphs['pos'] == '名詞'].index.tolist()
    no_idx = morphs[morphs['surface'] == 'の'].index.tolist()

    for idx in noun_idx:
        if idx+1 in no_idx and idx+2 in noun_idx:
            noun_no_noun.add(''.join(morphs.ix[[idx, idx+1, idx+2], :]['surface'].tolist()))


pprint(noun_no_noun)

File: scala/section4/src/main/scala/q34.scala
/**
  * Created by nomotoeriko on 2017/04/24.
  */
object q34 {

  def main(args: Array[String]): Unit = {
    val neko = q30.load_neko("../../data/neko.txt.mecab")
    val result = noun_no_noun(neko)
    result.foreach(println)
  }

  def noun_no_noun(neko: List[List[Map[String, String]]]): Set[String] = {
    neko.map(sentence => get_noun_no_noun(sentence)).reduce(_|_)
  }

  def get_noun_no_noun(sentence: List[Map[String, String]]): Set[String] = {
    val idx = (0 until sentence.length-2).toList.filter(i => sentence(i)("pos")=="名詞")
      .filter(i => sentence(i+1)("surface")=="の")
      .filter(i => sentence(i+2)("pos")=="名詞")
    idx.map(i => sentence(i)("surface")+sentence(i+1)("surface")+sentence(i+2)("surface")).toSet
  }
}
